{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Variant Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use this python script to connect to impala, find variants in subjects and gene regions of interest, annotate them with the following list of annotations, and save results to a file or impala table: \n",
    "\n",
    "- Kaviar allele frequency\n",
    "- ClinVar significance rating and disorder description\n",
    "- dbSNP rsID\n",
    "- DANN pathogenicity rating\n",
    "- Ensembl gene name and gene id\n",
    "- Predicted coding consequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup your parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each of the following parameters should be entered in python list format unless otherwise specified. Or enter 'all' to skip that particular filter. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# list of genes to search in\n",
    "gene_list = ['CYP%']\n",
    "\n",
    "# subject id's to include\n",
    "subject_list = ['102-00511-02', '5']\n",
    "#subject_list = 'all'\n",
    "\n",
    "# members to include (NB, M, F) or 'all'\n",
    "member_list = ['NB', 'M']\n",
    "#member_list = 'all'\n",
    "\n",
    "# genotype to include ('hom_rec', 'hom_alt', 'het') or 'all'\n",
    "geno_list = ['0/1']\n",
    "#geno_list = 'all'\n",
    "\n",
    "# max kaviar allele frequency to return, formatted as string, or 'all'\n",
    "max_kav_freq = '.03'\n",
    "\n",
    "# return only variants markes as non-conflicted significant in clinvar\n",
    "# (siginicance rating of 4 or 5, but never 2 or 3)\n",
    "# enter as 'yes' or 'no'\n",
    "clin_patho = 'yes'\n",
    "\n",
    "# minimum dann score to return\n",
    "min_dann = 'all'\n",
    "\n",
    "# return only variants with 'HIGH' coding consequences\n",
    "# enter 'yes' or 'no'\n",
    "coding_impact = 'no'\n",
    "\n",
    "# enter output method for results as 'tsv' or 'table'\n",
    "output_type = 'tsv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse User Arguments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create functions to parse user arguments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run this cell, but don't need to make any changes to it. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def label_member(trio_arg):\n",
    "    \"\"\"\n",
    "    function to create a sql statement from user trio argument for which\n",
    "    trio members to include in analysis\n",
    "    :param tbl_name: alias of tablename used in query as string, ex. 'bv'\n",
    "    :param trio_arg: list of trio members to include, ex. 'M','F','NB'\n",
    "    :return: member_arg\n",
    "    \"\"\"\n",
    "    member_list = []\n",
    "    for member in trio_arg:\n",
    "        if member == 'NB':\n",
    "            member_list.append(\"subject_id LIKE '%03'\")\n",
    "        if member == 'M':\n",
    "            member_list.append(\"subject_id LIKE '%01'\")\n",
    "        if member == 'F':\n",
    "            member_list.append(\"subject_id LIKE '%02'\")\n",
    "        if member_list == 'all':\n",
    "            member_list =''\n",
    "    if len(member_list) > 0:\n",
    "        member_arg = 'AND (' + ' OR '.join(member_list) + ')'\n",
    "    # otherwise statement is empty\n",
    "    else:\n",
    "        member_arg = ''\n",
    "    return member_arg\n",
    "\n",
    "def parse_arg_list(arg_list, col_name):\n",
    "    \"\"\"\n",
    "    format user arg lists \n",
    "    :param tbl_name: alias of tablename used in query as string, ex. 'vars'\n",
    "    :param arg_list: user arg list to parse\n",
    "    :return: query_arg assign to a variable to create sql statements\n",
    "    \"\"\"\n",
    "    wildcard_arg=[]\n",
    "    reg_arg = []\n",
    "    args = []\n",
    "    # if the user does not equal all\n",
    "    if arg_list != 'all':\n",
    "        # for each argument in the list\n",
    "        for arg in arg_list:\n",
    "            # if the argument contains a wildcard\n",
    "            if '%' in arg:\n",
    "                # append the argument to the wildcard list\n",
    "                wildcard_arg.append(arg)\n",
    "            # if no wildcard, append to regular arg list\n",
    "            else: \n",
    "                reg_arg.append(arg)\n",
    "    # if the user arg = 'all' create empty lists\n",
    "    else:\n",
    "        reg_arg = ''\n",
    "        wildcard_arg = ''\n",
    "    # create query statement from wildcard list\n",
    "    for arg in wildcard_arg:\n",
    "        # add a query statment to args for each arg\n",
    "        args.append(\"OR {} LIKE ('{}')\".format(col_name, arg))\n",
    "    for arg in reg_arg:\n",
    "        # add a query statment to args for each arg\n",
    "        args.append(\"OR {} = '{}'\".format(col_name, arg))\n",
    "    # final query statement joins all arguments in args\n",
    "    query_statement = ' '.join(args).replace('OR', 'AND (', 1)  \n",
    "    return query_statement \n",
    "    \n",
    "def parse_clinvar(clin_arg):\n",
    "    if clin_arg.lower() == 'yes':\n",
    "        clin_statement = \"AND (clin_sig NOT REGEXP '3|2[^5]|2$' AND clin_sig REGEXP '4|[^25]5|^5')\"\n",
    "    elif clin_arg.lower() == 'no':\n",
    "        clin_statement = ''\n",
    "    else:\n",
    "        print \"Enter a 'yes' or 'no' value for clin_patho.\"\n",
    "    return clin_statement\n",
    "\n",
    "def parse_coding(coding_arg):\n",
    "    if coding_arg.lower() == 'yes':\n",
    "        coding_statement = \"WHERE impact = 'HIGH'\"\n",
    "    elif coding_arg.lower() == 'no':\n",
    "        coding_statement = ''\n",
    "    else:\n",
    "        print \"Enter a 'yes' or 'no' value for coding_impact.\"\n",
    "    return coding_statement\n",
    "\n",
    "def parse_singles(user_arg, column, val_type, op_type):\n",
    "    if user_arg == 'all':\n",
    "        single_statement = ''\n",
    "    else:\n",
    "        if val_type == 'int':\n",
    "            single_statement = \"AND {} {} {}\".format(column, op_type, int(user_arg))\n",
    "        elif val_type == 'float':\n",
    "            single_statement = \"AND {} {} {}\".format(column, op_type, float(user_arg))\n",
    "        elif val_type == 'string':\n",
    "            single_statement = \"AND {} {} '{}'\".format(column, op_type, str(user_arg))\n",
    "        else:\n",
    "            print \"Please check that the data type of your {} argument is either int, float or string.\".format(user_arg)\n",
    "    return single_statement\n",
    "\n",
    "def parse_genotype(gt_arg):\n",
    "    gt_list = []\n",
    "    # if gt_arg = 'all' then leave blank\n",
    "    if gt_arg == 'all':\n",
    "        gt_statement = ''\n",
    "    else:\n",
    "        # if there is more than on gt arg\n",
    "        if len(gt_arg) > 1:\n",
    "            # add a statement to gt_list\n",
    "            gt_list.append(\"AND gt IN ('{}')\".format(\"','\".join(map(str, gt_arg))))\n",
    "        # if there is only one arg\n",
    "        elif len(gt_arg) == 1:\n",
    "            # add a gt '=' statement to the list\n",
    "            gt_list.append(\"AND gt = '{}'\".format(','.join(map(str, gt_arg))))\n",
    "    geno_statement = ' '.join(gt_list)\n",
    "    return geno_statement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "members = label_member(member_list)\n",
    "subjects = parse_arg_list(subject_list, 'subject_id')                                                \n",
    "genes = parse_arg_list(gene_list, 'ens_gene')\n",
    "geno = parse_genotype(geno_list)\n",
    "clin_statement = parse_clinvar(clin_patho)\n",
    "kav = parse_singles(max_kav_freq, 'kav_freq', 'float', '<=')\n",
    "dann = parse_singles(min_dann, 'dann_score', 'float', '>=')\n",
    "coding = parse_coding(coding_impact)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Merged Query Statements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "members = AND (subject_id LIKE '%03' OR subject_id LIKE '%01')\n",
      "subjects = AND ( subject_id = '102-00511-02' OR subject_id = '5'\n",
      "geno = AND gt = '0/1'\n",
      "WHERE (subject_id LIKE '%03' OR subject_id LIKE '%01') AND ( subject_id = '102-00511-02' OR subject_id = '5' AND gt = '0/1')\n"
     ]
    }
   ],
   "source": [
    "var_statement = (members + ' ' + subjects + ' ' + geno).replace('AND', 'WHERE', 1) \n",
    "if len(var_statement) > 0:\n",
    "    var_statement = var_statement + ')'\n",
    "\n",
    "print 'members = ' + members\n",
    "print 'subjects = ' + subjects \n",
    "print 'geno = ' + geno\n",
    "\n",
    "# # combine global variants query\n",
    "# gv_combo = [kav, clin_statement, dann]\n",
    "# gv_statement = ' '.join(gv_combo).replace('AND', 'WHERE (', 1) + ')'\n",
    "\n",
    "print var_statement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Write Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WITH vars AS \n",
      "    (\n",
      "    SELECT * \n",
      "    FROM p7_platform.wgs_illumina_variant \n",
      "         ))\n",
      "        AND gt = '0/1'\n",
      "     ),\n",
      "    gv as (\n",
      "        SELECT * \n",
      "        FROM p7_ref_grch37.global_variants\n",
      "        WHERE ( kav_freq <= 0.03 AND (clin_sig NOT REGEXP '3|2[^5]|2$' AND clin_sig REGEXP '4|[^25]5|^5') )\n",
      "        AND ( ens_gene LIKE ('CYP%'))),\n",
      "    cd as (\n",
      "        SELECT * \n",
      "        FROM users_selasady.coding_consequences\n",
      "        )\n",
      "     \n",
      "    SELECT vars.*, gv.kav_freq, gv.clin_sig, gv.clin_dbn, gv.rs_id,\n",
      "         gv.dann_score, gv.ens_gene, gv.ens_geneid, cd.effect, cd.impact,\n",
      "         cd.feature, cd.feature_id, cd.biotype, cd.rank, cd.hgvs_c, cd.hgvs_p\n",
      "         FROM vars, gv, cd\n",
      "         WHERE vars.chrom = gv.chrom\n",
      "         AND cd.chrom = gv.chrom\n",
      "         AND vars.pos = gv.pos\n",
      "         AND cd.pos = gv.pos\n",
      "         AND vars.ref = gv.ref\n",
      "         AND cd.ref = gv.ref\n",
      "         AND vars.alt = gv.alt\n",
      "         AND cd.alt = gv.alt\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "query = '''\n",
    "    WITH vars AS \n",
    "    (\n",
    "    SELECT * \n",
    "    FROM p7_platform.wgs_illumina_variant \n",
    "        {}\n",
    "        {}\n",
    "     ),\n",
    "    gv as (\n",
    "        SELECT * \n",
    "        FROM p7_ref_grch37.global_variants\n",
    "        {}\n",
    "        {}),\n",
    "    cd as (\n",
    "        SELECT * \n",
    "        FROM users_selasady.coding_consequences\n",
    "        {})\n",
    "     \n",
    "    SELECT vars.*, gv.kav_freq, gv.clin_sig, gv.clin_dbn, gv.rs_id,\n",
    "         gv.dann_score, gv.ens_gene, gv.ens_geneid, cd.effect, cd.impact,\n",
    "         cd.feature, cd.feature_id, cd.biotype, cd.rank, cd.hgvs_c, cd.hgvs_p\n",
    "         FROM vars, gv, cd\n",
    "         WHERE vars.chrom = gv.chrom\n",
    "         AND cd.chrom = gv.chrom\n",
    "         AND vars.pos = gv.pos\n",
    "         AND cd.pos = gv.pos\n",
    "         AND vars.ref = gv.ref\n",
    "         AND cd.ref = gv.ref\n",
    "         AND vars.alt = gv.alt\n",
    "         AND cd.alt = gv.alt\n",
    "    '''.format(subject_statement, geno, gv_statement, genes, coding)\n",
    "\n",
    "print query"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "##  Query Impala"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Connect to impala"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# import needed modules\n",
    "from impala.util import as_pandas\n",
    "import pandas as pd\n",
    "from impala.dbapi import connect\n",
    "\n",
    "# disable extraneous pandas warning\n",
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def run_query(query_name, out_db, out_name):\n",
    "    \"\"\"\n",
    "    opens odbc connection to impala, drops table if exists, runs query\n",
    "     saves results as table, closes connection\n",
    "    :param query_name: name of query statement to run\n",
    "    :param out_db: name of output database\n",
    "    :param out_name: name of output table name\n",
    "    :return: table of results saved on impala in specified output dir\n",
    "    \"\"\"\n",
    "    # create connection object\n",
    "    conn=connect(host='glados19', port=21050, timeout=120)\n",
    "    # drop table if it exists\n",
    "    cur = conn.cursor()\n",
    "    print 'Removing table if it already exists...'\n",
    "    cur.execute('DROP TABLE IF EXISTS {}.{}'.format(out_db, out_name))\n",
    "    # run query\n",
    "    print 'Running the following query on impala: \\n' + query_name\n",
    "    cur.execute(query_name)\n",
    "    results_df = as_pandas(cur)\n",
    "    if len(results_df) > 0:\n",
    "        print 'Query finished. Closing connection.'\n",
    "        return results_df\n",
    "    else:\n",
    "        print 'No results found.'\n",
    "    cur.close()\n",
    "    conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing table if it already exists...\n",
      "Running the following query on impala: \n",
      "\n",
      "    WITH vars AS \n",
      "    (\n",
      "    SELECT * \n",
      "    FROM p7_platform.wgs_illumina_variant \n",
      "         ))\n",
      "        AND gt = '0/1'\n",
      "     ),\n",
      "    gv as (\n",
      "        SELECT * \n",
      "        FROM p7_ref_grch37.global_variants\n",
      "        WHERE ( kav_freq <= 0.03 AND (clin_sig NOT REGEXP '3|2[^5]|2$' AND clin_sig REGEXP '4|[^25]5|^5') )\n",
      "        AND ( ens_gene LIKE ('CYP%'))),\n",
      "    cd as (\n",
      "        SELECT * \n",
      "        FROM users_selasady.coding_consequences\n",
      "        )\n",
      "     \n",
      "    SELECT vars.*, gv.kav_freq, gv.clin_sig, gv.clin_dbn, gv.rs_id,\n",
      "         gv.dann_score, gv.ens_gene, gv.ens_geneid, cd.effect, cd.impact,\n",
      "         cd.feature, cd.feature_id, cd.biotype, cd.rank, cd.hgvs_c, cd.hgvs_p\n",
      "         FROM vars, gv, cd\n",
      "         WHERE vars.chrom = gv.chrom\n",
      "         AND cd.chrom = gv.chrom\n",
      "         AND vars.pos = gv.pos\n",
      "         AND cd.pos = gv.pos\n",
      "         AND vars.ref = gv.ref\n",
      "         AND cd.ref = gv.ref\n",
      "         AND vars.alt = gv.alt\n",
      "         AND cd.alt = gv.alt\n",
      "    \n"
     ]
    },
    {
     "ename": "HiveServer2Error",
     "evalue": "AnalysisException: Syntax error in line 6:\n         ))\n          ^\nEncountered: )\nExpected: INSERT, SELECT, VALUES, COMMA\n\nCAUSED BY: Exception: Syntax error",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mHiveServer2Error\u001b[0m                          Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-180-5babb5fe364b>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m# run query\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mresults_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrun_query\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'users_selasady'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'filtered_variants'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-179-4bba12652b8c>\u001b[0m in \u001b[0;36mrun_query\u001b[1;34m(query_name, out_db, out_name)\u001b[0m\n\u001b[0;32m     16\u001b[0m     \u001b[1;31m# run query\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     17\u001b[0m     \u001b[1;32mprint\u001b[0m \u001b[1;34m'Running the following query on impala: \\n'\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mquery_name\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 18\u001b[1;33m     \u001b[0mcur\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mquery_name\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     19\u001b[0m     \u001b[0mresults_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mas_pandas\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcur\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     20\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresults_df\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\Users\\Summer\\AppData\\Roaming\\Python\\Python27\\site-packages\\impala\\hiveserver2.pyc\u001b[0m in \u001b[0;36mexecute\u001b[1;34m(self, operation, parameters, configuration)\u001b[0m\n\u001b[0;32m    185\u001b[0m         \u001b[1;31m# PEP 249\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    186\u001b[0m         self.execute_async(operation, parameters=parameters,\n\u001b[1;32m--> 187\u001b[1;33m                            configuration=configuration)\n\u001b[0m\u001b[0;32m    188\u001b[0m         \u001b[0mlog\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Waiting for query to finish'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    189\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_wait_to_finish\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m  \u001b[1;31m# make execute synchronous\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\Users\\Summer\\AppData\\Roaming\\Python\\Python27\\site-packages\\impala\\hiveserver2.pyc\u001b[0m in \u001b[0;36mexecute_async\u001b[1;34m(self, operation, parameters, configuration)\u001b[0m\n\u001b[0;32m    203\u001b[0m                 configuration)\n\u001b[0;32m    204\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 205\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_execute_async\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mop\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    206\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    207\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_debug_log_state\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\Users\\Summer\\AppData\\Roaming\\Python\\Python27\\site-packages\\impala\\hiveserver2.pyc\u001b[0m in \u001b[0;36m_execute_async\u001b[1;34m(self, operation_fn)\u001b[0m\n\u001b[0;32m    218\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_reset_state\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    219\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_debug_log_state\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 220\u001b[1;33m         \u001b[0moperation_fn\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    221\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_last_operation_active\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTrue\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    222\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_debug_log_state\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\Users\\Summer\\AppData\\Roaming\\Python\\Python27\\site-packages\\impala\\hiveserver2.pyc\u001b[0m in \u001b[0;36mop\u001b[1;34m()\u001b[0m\n\u001b[0;32m    201\u001b[0m             self._last_operation_handle = execute_statement(\n\u001b[0;32m    202\u001b[0m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mservice\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msession_handle\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_last_operation_string\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 203\u001b[1;33m                 configuration)\n\u001b[0m\u001b[0;32m    204\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    205\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_execute_async\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mop\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\Users\\Summer\\AppData\\Roaming\\Python\\Python27\\site-packages\\impala\\hiveserver2.pyc\u001b[0m in \u001b[0;36mwrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    555\u001b[0m                     \u001b[0mtransport\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    556\u001b[0m                 \u001b[0mlog\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Transport opened'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 557\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    558\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0msocket\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0merror\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    559\u001b[0m                 log.exception('Failed to open transport (tries_left=%s)',\n",
      "\u001b[1;32mC:\\Users\\Summer\\AppData\\Roaming\\Python\\Python27\\site-packages\\impala\\hiveserver2.pyc\u001b[0m in \u001b[0;36mexecute_statement\u001b[1;34m(service, session_handle, statement, configuration, async)\u001b[0m\n\u001b[0;32m    640\u001b[0m     \u001b[0mresp\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mservice\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mExecuteStatement\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    641\u001b[0m     \u001b[0mlog\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'execute_statement: resp=%s'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mresp\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 642\u001b[1;33m     \u001b[0merr_if_rpc_not_ok\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresp\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    643\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moperationHandle\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    644\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\Users\\Summer\\AppData\\Roaming\\Python\\Python27\\site-packages\\impala\\hiveserver2.pyc\u001b[0m in \u001b[0;36merr_if_rpc_not_ok\u001b[1;34m(resp)\u001b[0m\n\u001b[0;32m    495\u001b[0m             \u001b[0mresp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstatus\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstatusCode\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[0mTStatusCode\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSUCCESS_WITH_INFO_STATUS\u001b[0m \u001b[1;32mand\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    496\u001b[0m             resp.status.statusCode != TStatusCode.STILL_EXECUTING_STATUS):\n\u001b[1;32m--> 497\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mHiveServer2Error\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstatus\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0merrorMessage\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    498\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    499\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mHiveServer2Error\u001b[0m: AnalysisException: Syntax error in line 6:\n         ))\n          ^\nEncountered: )\nExpected: INSERT, SELECT, VALUES, COMMA\n\nCAUSED BY: Exception: Syntax error"
     ]
    }
   ],
   "source": [
    "# run query\n",
    "results_df = run_query(query, 'users_selasady', 'filtered_variants')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'type'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-165-874e79ee06a5>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;32mprint\u001b[0m \u001b[0mresults_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'type'"
     ]
    }
   ],
   "source": [
    "print results_df.type()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
